packages/runtime/src/next/api-handler/handlers/webhook/index.ts

```diff
@@ -0,0 +1,75 @@
+import { NextApiRequest, NextApiResponse } from 'next'
+import { NextRequest, NextResponse } from 'next/server'
+import { P, match } from 'ts-pattern'
+import { handleSitePublished } from './site-published'
+import {
+  sitePublishedWebhookPayloadSchema,
+  WebhookEventType,
+  WebhookPayloadSchema,
+  WebhookResponseBody,
+} from './types'
+
+type Context = { params: { [key: string]: string | string[] } }
+
+export type WebhookHandlerArgs =
+  | [request: NextRequest, context: Context, params: { apiKey: string }]
+  | [req: NextApiRequest, res: NextApiResponse<WebhookResponseBody>, params: { apiKey: string }]
+
+const routeHandlerPattern = [P.instanceOf(Request), P.any, P.any] as const
+const apiRoutePattern = [P.any, P.any, P.any] as const
+
+function getSecret(args: WebhookHandlerArgs) {
+  return match(args)
+    .with(routeHandlerPattern, ([request]) => request.nextUrl.searchParams.get('secret'))
+    .with(apiRoutePattern, ([req]) => req.query.secret)
+    .exhaustive()
+}
+
+function getRequestBody(args: WebhookHandlerArgs) {
+  return match(args)
+    .with(routeHandlerPattern, ([request]) => request.json())
+    .with(apiRoutePattern, ([req]) => req.body)
+    .exhaustive()
+}
+
+function respond(args: WebhookHandlerArgs, response: WebhookResponseBody, status?: number) {
+  return match(args)
+    .with(routeHandlerPattern, () => NextResponse.json(response, { status }))
+    .with(apiRoutePattern, ([, res]) => res.json(response))
+    .exhaustive()
+}
+
+export default async function handler(
+  request: NextRequest,
+  context: Context,
+  { apiKey }: { apiKey: string },
+): Promise<NextResponse<WebhookResponseBody>>
+export default async function handler(
+  req: NextApiRequest,
+  res: NextApiResponse<WebhookResponseBody>,
+  { apiKey }: { apiKey: string },
+): Promise<void>
+export default async function handler(
+  ...args: WebhookHandlerArgs
+): Promise<NextResponse<WebhookResponseBody> | void> {
+  const [, , { apiKey }] = args
+  const secret = getSecret(args)
+
+  if (secret !== apiKey) return respond(args, { message: 'Unauthorized' }, 401)
+
+  let payload: WebhookPayloadSchema
+
+  try {
+    const body = await getRequestBody(args)
+    payload = sitePublishedWebhookPayloadSchema.parse(body)
```

Comment:
What's the status on us adding web hook signatures? If we have support for them already, we should validate the signature here

Reply:
We don't have webhook signatures right now. We validate the request by the `siteApiKey`. Is this enough?

Reply:
I guess we're using the site's API key as the signature. It feels weird to me that we're using the site's API key this way instead of having a web hook secret that can be used to create a signature to ensure the web hook payload is to be trusted. Sending the API key on each request (which I know we already do for revalidation) feels very insecure. Maybe there's nothing to do right now since we're using the site's API key for revalidation and we haven't implemented web hook signatures just yet but I feel dirty letting this slide...

<details>
<summary>Here's what the oracle says</summary>

When implementing webhooks, using a **webhook signature** is generally preferred over sending a **shared API key** in the URL or headers for several important security reasons:

### 1. **Reduced Risk of Exposure**
   - **API key exposure**: Sending an API key in the URL or headers makes it vulnerable to interception. URLs (including query parameters) may be logged in web server logs, browser history, or through third-party monitoring tools. This increases the risk of exposing sensitive credentials.
   - **Signature approach**: With a webhook signature, you are sending a hashed value that is generated based on the payload and a secret key. Even if the signature is intercepted, without the secret key, it cannot be verified or manipulated to produce a valid signature. This adds a layer of security by not directly exposing any sensitive data in transit.

### 2. **Integrity and Authenticity**
   - **API key validation**: API keys can be easily guessed or copied if not stored securely. An attacker who gains access to the key can impersonate the webhook provider, sending fraudulent requests.
   - **Signature validation**: Webhook signatures use cryptographic hashing (usually HMAC or similar) to ensure that the payload has not been tampered with in transit. The receiving server checks the signature against the payload and a shared secret to verify that the request genuinely comes from the expected source and that the data hasn’t been altered.

### 3. **Better Protection Against Replay Attacks**
   - **API key exposure**: If an API key is used in the query string or headers, it could potentially be reused by an attacker (replay attack), allowing them to resend requests from valid sources to the server without needing to authenticate or sign the request.
   - **Signature approach**: Signatures typically incorporate time-based or nonce-based components to prevent replay attacks. For example, a webhook signature could include a timestamp, and if a request is received with an outdated timestamp, it can be rejected as a potential replay attack.

### 4. **No Need for Storing Sensitive Data in Requests**
   - **API keys in headers/URL**: Storing sensitive keys in requests, especially in plain text or query parameters, increases the risk of unintentional exposure through logging or inspection tools.
   - **Signature approach**: The signature only reveals the result of a hashed process, not the original secret key. As long as the secret key is kept secure on the server and is not transmitted in the request, the attack surface is significantly reduced.

### 5. **Flexible Security Handling**
   - **API keys**: With API keys, any change (e.g., key rotation) would require updates to all clients, which can be cumbersome and error-prone.
   - **Signature approach**: If a secret key needs to be rotated, the update is often more straightforward since the signature is generated from a known shared secret but doesn't expose the key itself in the request. The system can easily re-sign with the new key without altering the rest of the request structure.

### 6. **Prevents Unauthorized Access**
   - **API key-based access**: Once an API key is exposed or leaked, attackers could potentially gain access to the service and impersonate the webhook provider.
   - **Signature-based security**: Since the signature is unique to the specific payload and time, unauthorized parties cannot forge the correct signature without access to the secret key. This makes it far more difficult for attackers to gain unauthorized access.

### Summary
In short, **webhook signatures** offer a more secure method for verifying the authenticity of webhook requests, ensuring the integrity of the payload, and minimizing the risk of exposure or misuse of sensitive credentials like API keys. They provide an added layer of security by relying on cryptographic signatures instead of transmitting shared secrets directly in the request.
</details>

Reply:
I understand. We need to do the same for all the API handlers like the manifest too. We decided not to worry about this for now.